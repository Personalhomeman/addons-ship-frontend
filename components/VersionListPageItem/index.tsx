import { Component } from 'react';
import css from './style.scss';
import SVG from 'react-svg';
import Link from 'next/link';

interface VersionListPageItemProps {
  detailsPagePath: string;
  platformIconUrl: string;
  title: string;
  description: string;
  note: string;
}

type VersionListPageItemState = {
  isOpen: boolean;
};

export default class VersionListPageItem extends Component<VersionListPageItemProps, VersionListPageItemState> {
  state: VersionListPageItemState = {
    isOpen: false
  };

  toggle = () => {
    const { isOpen } = this.state;
    this.setState({ isOpen: !isOpen });
  };

  render() {
    const { detailsPagePath, platformIconUrl, title, description, note } = this.props;
    const { isOpen } = this.state;

    const descriptionWrapperClasses = [css.descriptionWrapper];
    if (isOpen) {
      descriptionWrapperClasses.push(css.isOpen);
    }

    return (
      <div className={css.versionListPageItem}>
        <div className={css.versionListPageItemInner}>
          <Link href={detailsPagePath}>
            <a className={css.topWrapper}>
              <SVG src={platformIconUrl} className={css.platformIcon} />
              <div className={css.title}>{title}</div>
              <div className={css.note}>{note}</div>
            </a>
          </Link>
          <button className={descriptionWrapperClasses.join(' ')} onClick={this.toggle}>
            <div className={css.description}>{description}</div>
            {isOpen ? (
              <div className={css.showLess}>
                <div className={css.text}>Show less</div>
                <SVG src="/static/arrow-down.svg" className={css.arrow} />
              </div>
            ) : (
              <div className={css.showMore}>
                <div className={css.text}>Show more</div>
                <SVG src="/static/arrow-down.svg" className={css.arrow} />
              </div>
            )}
          </button>
        </div>
      </div>
    );
  }
}
