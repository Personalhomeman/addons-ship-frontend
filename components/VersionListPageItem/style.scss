@import '~@/assets/style/mixins';
@import '~@/assets/style/dimensions';
@import '~@/assets/style/colors';
@import '~@/assets/style/times';

.versionListPageItem {
  @include padding-y($grid);
  border-bottom: 1px solid $gray-2;

  .versionListPageItemInner {
    border-radius: 2 * $grid;
    background-color: white;
    transition: background-color $medium-hovertime;

    .topWrapper {
      display: flex;
      align-items: center;
      padding-top: 4 * $grid;
      padding-bottom: 2 * $grid;
      @include padding-x(4 * $grid);
      text-decoration: none;

      .platformIcon {
        margin-right: $grid;
        margin-bottom: 3px;

        svg {
          display: block;
          fill: $grape-5;
          transition: fill $medium-hovertime;
        }
      }

      .title {
        flex-grow: 1;
        color: $grape-5;
        transition: color $medium-hovertime;
        // TODO: replace with element from UI library
        line-height: 1.5;
        letter-spacing: normal;
        font-style: normal;
        font-stretch: normal;
        font-size: 19px;
        font-weight: 700;
      }

      .note {
        color: $gray-6;
        transition: color $medium-hovertime;
        // TODO: replace with element from UI library
        font-size: 16px;
        line-height: 1.5;
        letter-spacing: normal;
        font-weight: 500;
        font-style: normal;
        font-stretch: normal;
      }
    }

    .descriptionWrapper {
      display: flex;
      align-items: center;
      width: 100%;
      border: none;
      padding-top: 2 * $grid;
      padding-bottom: 4 * $grid;
      @include padding-x(4 * $grid);
      text-align: left;
      font-family: inherit;
      background-color: transparent;
      cursor: pointer;

      .description {
        flex-grow: 1;
        color: $grape-5;
        transition: color $medium-hovertime;
        // TODO: replace with element from UI library
        font-size: 16px;
        line-height: 1.5;
        letter-spacing: normal;
        font-weight: 500;
        font-style: normal;
        font-stretch: normal;
      }

      &:not(.isOpen) {
        .description {
          @include ellipsis-oneliner;
        }

        .showMore {
          display: none;
        }
      }

      &.isOpen {
        display: block;

        .showLess {
          display: flex;
          align-items: center;
          margin-top: 2 * $grid;

          .text {
            color: $grape-5;
            // TODO: replace with element from UI library
            font-size: 13px;
            line-height: 4 * $grid;
            letter-spacing: normal;
            font-weight: 500;
            font-style: normal;
            font-stretch: normal;
          }

          .arrow {
            flex-shrink: 0;

            svg {
              display: block;
              @include size(4 * $grid);
              transform: rotate(180deg);
              fill: $grape-5;
            }
          }
        }
      }
    }

    &:hover,
    &:focus {
      background-color: $gray-1;

      .topWrapper {
        .platformIcon {
          svg {
            fill: $grape-3;
          }
        }

        .title {
          color: $grape-3;
        }

        .note {
          color: $grape-3;
        }
      }

      .descriptionWrapper {
        .description {
          color: $grape-3;
        }

        &:not(.isOpen) .showMore {
          flex-shrink: 0;
          display: flex;
          align-items: center;
          margin-left: 2 * $grid;

          .text {
            color: $grape-3;
            // TODO: replace with element from UI library
            font-size: 13px;
            line-height: 4 * $grid;
            letter-spacing: normal;
            font-weight: 500;
            font-style: normal;
            font-stretch: normal;
          }

          .arrow {
            flex-shrink: 0;

            svg {
              display: block;
              @include size(4 * $grid);
              fill: $grape-3;
            }
          }
        }

        &.isOpen {
          .showLess {
            .text {
              color: $grape-3;
            }

            .arrow {
              svg {
                fill: $grape-3;
              }
            }
          }
        }
      }
    }
  }
}
