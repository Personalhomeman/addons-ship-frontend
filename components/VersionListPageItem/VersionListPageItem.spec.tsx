import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import VersionListPageItem from '.';

describe('VersionListPageItem', () => {
  it('renders correctly', () => {
    const component = shallow(
      <VersionListPageItem
        detailsPagePath="/v1-1-0/details"
        detailsPagePathHref="/v1-1-0/details"
        platform="ios"
        title="My app v1.1.0 (28)"
        description="This is a short description about my 1. app. It is a good app. End of description."
        note="Updated on January 1, 2008"
      />
    );
    expect(toJSON(component)).toMatchSnapshot();
  });

  it('renders correctly for android', () => {
    const component = shallow(
      <VersionListPageItem
        detailsPagePath="/v1-1-0/details"
        detailsPagePathHref="/v1-1-0/details"
        platform="android"
        title="My app v1.1.0 (28)"
        description="This is a short description about my 1. app. It is a good app. End of description."
        note="Updated on January 1, 2008"
        productFlavor="banana"
      />
    );
    expect(toJSON(component)).toMatchSnapshot();
  });

  it('shows more description', () => {
    const component = shallow(
      <VersionListPageItem
        detailsPagePath="/v1-1-0/details"
        detailsPagePathHref="/v1-1-0/details"
        platform="ios"
        title="My app v1.1.0 (28)"
        description="This is a short description about my 1. app. It is a good app. End of description."
        note="Updated on January 1, 2008"
      />
    );

    const button = component.find('.descriptionWrapper');
    button.simulate('click');

    expect(toJSON(component)).toMatchSnapshot();
  });
});
