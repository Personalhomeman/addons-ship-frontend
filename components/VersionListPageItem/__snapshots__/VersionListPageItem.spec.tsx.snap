// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VersionListPageItem renders correctly 1`] = `
<div
  className="versionListPageItem"
>
  <div
    className="versionListPageItemInner"
  >
    <Link
      href="/v1-1-0/details"
    >
      <a
        className="topWrapper"
      >
        <ReactSVG
          afterInjection={[Function]}
          beforeInjection={[Function]}
          className="platformIcon"
          evalScripts="never"
          fallback={null}
          loading={null}
          renumerateIRIElements={true}
          src="/static/icon-apple.svg"
          wrapper="div"
        />
        <div
          className="title"
        >
          My app v1.1.0 (28)
        </div>
        <div
          className="note"
        >
          Updated on January 1, 2008
        </div>
      </a>
    </Link>
    <button
      className="descriptionWrapper"
      onClick={[Function]}
    >
      <div
        className="description"
      >
        This is a short description about my 1. app. It is a good app. End of description.
      </div>
      <div
        className="showMore"
      >
        <div
          className="text"
        >
          Show more
        </div>
        <ReactSVG
          afterInjection={[Function]}
          beforeInjection={[Function]}
          className="arrow"
          evalScripts="never"
          fallback={null}
          loading={null}
          renumerateIRIElements={true}
          src="/static/arrow-down.svg"
          wrapper="div"
        />
      </div>
    </button>
  </div>
</div>
`;

exports[`VersionListPageItem shows more description 1`] = `
<div
  className="versionListPageItem"
>
  <div
    className="versionListPageItemInner"
  >
    <Link
      href="/v1-1-0/details"
    >
      <a
        className="topWrapper"
      >
        <ReactSVG
          afterInjection={[Function]}
          beforeInjection={[Function]}
          className="platformIcon"
          evalScripts="never"
          fallback={null}
          loading={null}
          renumerateIRIElements={true}
          src="/static/icon-apple.svg"
          wrapper="div"
        />
        <div
          className="title"
        >
          My app v1.1.0 (28)
        </div>
        <div
          className="note"
        >
          Updated on January 1, 2008
        </div>
      </a>
    </Link>
    <button
      className="descriptionWrapper isOpen"
      onClick={[Function]}
    >
      <div
        className="description"
      >
        This is a short description about my 1. app. It is a good app. End of description.
      </div>
      <div
        className="showLess"
      >
        <div
          className="text"
        >
          Show less
        </div>
        <ReactSVG
          afterInjection={[Function]}
          beforeInjection={[Function]}
          className="arrow"
          evalScripts="never"
          fallback={null}
          loading={null}
          renumerateIRIElements={true}
          src="/static/arrow-down.svg"
          wrapper="div"
        />
      </div>
    </button>
  </div>
</div>
`;
